### PR TITLE
Devops/prod docker

### PR DESCRIPTION
**Pull request by Meital-Kozhidov**
docker-compose & additions to dockerfiles to visualize production.

Run with:
`DB_INIT_SCRIPT=../be/db/migrations/0000_wealthy_miss_america.sql docker-compose up --build -d --wait --force-recreate`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,9 +1,11 @@
+################################ Development ###################################
+
 FROM node:20.11.1-alpine as dev-base
 
 # Install dependencies
 RUN apk update && apk add --no-cache curl=8.5.0-r0 && apk add --no-cache tini=0.19.0-r2
 
-############################### Development BE ##################################
+################################## BE ##########################################
```
Comment (AdamAkiva) [inline Dockerfile:8]: I respect this, don't agree, but respect

File: .gitignore [hunk 2]
```diff
@@ -34,6 +34,8 @@
 !*.sh
 !*.md
 !*.yml
+!proxy.conf
+!*.conf.template
```
Comment (AdamAkiva) [inline .gitignore:38]: Why `*.conf.template` instead of `dashboard.conf.template`?
Is it not better to have a more explicit approach?

File: Dockerfile [hunk 3]
```diff
@@ -19,7 +21,13 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
 # Run the script as PID 1
 CMD ["/home/node/entrypoint.sh"]
 
-############################### Development FE ##################################
+FROM node:20.11.1-alpine as be-base
```
Comment (AdamAkiva) [inline Dockerfile:24]: If you called it below `prod base`, should this not be `dev-base`?

File: Dockerfile [hunk 4]
```diff
@@ -34,3 +42,83 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
...
+# Copy build resources and source code
+COPY be/eslint.config.js be/package*.json be/tsconfig*.json ./
+COPY be/src ./src
+
+# Build for production
+RUN npm ci
+RUN npm run _build
+RUN npm prune --production
+
+# Create folder for only the necesseties for serving
```
Comment (AdamAkiva) [inline Dockerfile:70]: necesseties -> necessities

File: Dockerfile [hunk 5]
```diff
@@ -34,3 +42,83 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
...
+# And copy the necesseties to the folder
+RUN /bin/sh -c "mkdir -p prod_build && cp -r ./build/* ./package.json ./node_modules ./prod_build"
```
Comment (AdamAkiva) [inline Dockerfile:72]: After some research about `tini` and the signals which showed me stuff regrading the docker shell, I think it's better to separate it into two different `RUN` commands

File: Dockerfile [hunk 6]
```diff
@@ -34,3 +42,83 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
...
+
+FROM prod-base as be-serve
+
+# Set the workdir
+WORKDIR /home/node/dashboard
+
+# Copy build folder from previous stage with node priviliges
+COPY --chown=node:node --from=be-build /home/node/dashboard/prod_build ./
+
+ENTRYPOINT ["/bin/sh", "-c", "NODE_ENV=production node ./main.js"]
```
Comment (AdamAkiva) [inline Dockerfile:85]: This will not work with piping signals.
Should be changed to:
`ENTRYPOINT ["NODE_ENV = production node", "./main.js"]`
Note: I didn't check it works, but the point is to remove the `/bin/sh -c` since it spawns a new shell

File: Dockerfile [hunk 7]
```diff
@@ -34,3 +42,83 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
...
+COPY --chown=nginx:nginx fe/nginx /etc/nginx
+
+# Copy build folder from previous stage with nginx priviliges
+COPY --chown=nginx:nginx --from=fe-build /home/node/dashboard/build ./
+
+# CMD is used in contrast to ENTRYPOINT because the nginx docker image uses
+# ENTRYPOINT to run the scripts located in /docker-entrypoint.d
+CMD ["nginx", "-g", "daemon off;"]
+
+# Nginx runs on port 8080 by default
```
Comment (AdamAkiva) [inline Dockerfile:123]: Untrue, NGINX runs on 80 by default:
https://www.google.com/search?channel=fs&client=ubuntu&q=nginx+default+port

File: devops/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  ##############################################################################
+
+  dashboard-fe-prod:
+    container_name: dashboard-fe-prod
+    build:
+      context: ../
+      dockerfile: Dockerfile
+      target: fe-serve
+      args:
+        - DASHBOARD_SERVER_API_URL=http://dashboard.local.co.il/v0/api
```
Comment (AdamAkiva) [inline devops/docker-compose.yml:92]: Maybe add a note about the need for the /etc/hosts addition?

File: devops/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    environment:
+        - NGINX_SERVER_URL=dashboard-be-prod:2828
```
Comment (AdamAkiva) [inline devops/docker-compose.yml:94]: Maybe add a note here about why this is the URL?
(I know it is written in the NGINX file, but most people will not know where to look for it)

File: fe/nginx/proxy.conf [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# Proxy configuration (prevent code bloat in dashboard.conf.template)
```
Comment (AdamAkiva) [inline fe/nginx/proxy.conf:1]: The comment is correct, but the purpose is to allow web socket connections, consider adding that.
(Yes I know it is copied from amey but I was much less knowledgeable then)

File: fe/nginx/templates/dashboard.conf.template [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    server ${NGINX_SERVER_URL};
+}
+
+# Dashboard static serving and reverse proxy server
+server {
+    listen ${NGINX_PORT};
+    server_name ${DASHBOARD_URL};
+
+    # Path inside the nginx docker
+    root /usr/share/nginx/html;
```
Comment (AdamAkiva) [inline fe/nginx/templates/dashboard.conf.template:14]: Not true, that is not the path the the FE was copied to inside the docker
See: http://nginx.org/en/docs/http/ngx_http_core_module.html#root

File: fe/nginx/templates/dashboard.conf.template [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+    # Reverse proxy to the backend including web sockets support
```
Comment (AdamAkiva) [inline fe/nginx/templates/dashboard.conf.template:16]: Maybe (including web sockets support)?
Adding parenthesis

File: fe/src/utils/env.ts
Comment (AdamAkiva) [inline fe/src/utils/env.ts:1]: Why did you remove this?
Looks the same usage as the what you did in the `api.ts` file?
What am I missing?

File: fe/src/api/api.ts [hunk 13]
```diff
@@ -22,7 +22,7 @@ import { Http } from './httpClient';
 /******************************************************************************/
 
 const http = new Http({
-  prefixUrl: getEnvValue('SERVER_API_URL'),
+  prefixUrl: import.meta.env.DASHBOARD_SERVER_API_URL,
```
Comment (AdamAkiva) [inline fe/src/api/api.ts:25]: Why the change?

Comment (Meital-Kozhidov) [reply fe/src/api/api.ts:25]: from  https://vitejs.dev/guide/build.html

```The exception is when you need to dynamically concatenate URLs on the fly. In this case, you can use the globally injected import.meta.env.BASE_URL variable which will be the public base path. Note this variable is statically replaced during build so it must appear exactly as-is (i.e. import.meta.env['BASE_URL'] won't work).
```

File: fe/src/utils/env.ts
Comment (Meital-Kozhidov) [reply fe/src/utils/env.ts:1]: See answer below

File: Dockerfile (hunk 4 above)
Comment (Meital-Kozhidov) [reply Dockerfile:70]: fixed

File: Dockerfile (hunk 5 above)
Comment (Meital-Kozhidov) [reply Dockerfile:72]: fixed

File: Dockerfile (hunk 6 above)
Comment (Meital-Kozhidov) [reply Dockerfile:85]: fixed

File: Dockerfile (hunk 7 above)
Comment (Meital-Kozhidov) [reply Dockerfile:123]: fixed

File: fe/nginx/proxy.conf (hunk 10 above)
Comment (Meital-Kozhidov) [reply fe/nginx/proxy.conf:1]: added

File: fe/nginx/templates/dashboard.conf.template (hunk 12 above)
Comment (Meital-Kozhidov) [reply fe/nginx/templates/dashboard.conf.template:16]: added

File: fe/src/api/api.ts (hunk 13 above)
Comment (Meital-Kozhidov) [reply fe/src/api/api.ts:25]: Or am i stupid and this is just about BASE_URL? it didn't work for me so I thought it was that, but I changed things since. Will check again and if it works I'll restore the env file

File: fe/nginx/templates/dashboard.conf.template (hunk 11 above)
Comment (Meital-Kozhidov) [reply fe/nginx/templates/dashboard.conf.template:14]: Where should i look in the link

File: Dockerfile [hunk 14]
```diff
@@ -34,3 +37,84 @@ ENTRYPOINT ["/sbin/tini", "-s", "--"]
...
+COPY fe/src ./src
+
+# Build for production
+ARG DASHBOARD_SERVER_API_URL=${DASHBOARD_SERVER_API_URL}
+RUN npm ci
+RUN npm run _build
+
+################################# Serve FE #####################################
+
+# Unpriviliged - to not have root priviliges
```
Comment (AdamAkiva) [inline Dockerfile:102]: Unpriviliged -> Unprivileged
priviliges -> privileges
Have a look at [this](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)
If you have it or something like it maybe the configurations are not correct?

File: fe/nginx/templates/dashboard.conf.template (hunk 11 above)
Comment (AdamAkiva) [reply fe/nginx/templates/dashboard.conf.template:14]: It is directed at the directive which points to the wrong location.
The `root` directive should point to build folder.
image

File: fe/src/api/api.ts (hunk 13 above)
Comment (AdamAkiva) [reply fe/src/api/api.ts:25]: From the link you provided I understood that it refers specially to `BASE_URL` variable of `vite`. (The one shown [here](https://vitejs.dev/guide/env-and-mode.html))
If and when you get the chance I'd love an update.

File: Dockerfile (hunk 6 above)
Comment (AdamAkiva) [reply Dockerfile:81]: Then do this instead:
```docker
FROM prod-base as be-serve

ENV NODE_ENV=production

# Set the workdir
WORKDIR /home/node/dashboard

# Copy build folder from previous stage with node privileges
COPY --chown=node:node --from=be-build /home/node/dashboard/prod_build ./

ENTRYPOINT ["node", "./main.js"]
```
I've checked, it works

File: fe/nginx/templates/dashboard.conf.template (hunk 11 above)
Comment (AdamAkiva) [reply fe/nginx/templates/dashboard.conf.template:14]: Exactly where it point, image attached for reference.
(At least that's what I understand from it, however since it works maybe I don't understand anything)
image

File: Dockerfile (hunk 14 above)
Comment (Meital-Kozhidov) [reply Dockerfile:102]: I have this extension and it works, but not in dockerfiles. I understand it does work for you?

File: fe/src/api/api.ts (hunk 13 above)
Comment (Meital-Kozhidov) [reply fe/src/api/api.ts:25]: You're correct. I restored the env file. Thanks

File: Dockerfile (hunk 6 above)
Comment (Meital-Kozhidov) [reply Dockerfile:81]: Done

File: devops/docker-compose.yml (hunk 8 above)
Comment (Meital-Kozhidov) [reply devops/docker-compose.yml:92]: Done

File: devops/docker-compose.yml (hunk 9 above)
Comment (Meital-Kozhidov) [reply devops/docker-compose.yml:94]: Done

File: fe/nginx/templates/dashboard.conf.template (hunk 11 above)
Comment (Meital-Kozhidov) [reply fe/nginx/templates/dashboard.conf.template:14]: image
Am i missing something? does seem the correct path inside the docker. Do you want me to change the comment to 'sets root directory for requests'?

File: Dockerfile (hunk 14 above)
Comment (AdamAkiva) [reply Dockerfile]: Yes, the settings I use are:
```json
"cSpell.enableFiletypes": ["*"],
"cSpell.language": "en",
```
In my `settings.json` file

File: devops/docker-compose.yml [hunk 15]
```diff
@@ -89,8 +89,10 @@ services:
       dockerfile: Dockerfile
       target: fe-serve
       args:
+        # Requires mapping in /etc/hosts from 'dashboard.local.co.il' to 'localhost'
```
Comment (AdamAkiva) [inline devops/docker-compose.yml:92]: Should it not be the other way?
Requires mapping on /etc/hosts file from 'localhost' (or 127.0.0.1) to 'dashboard.local.co.il'?

File: Dockerfile (hunk 14 above)
Comment (AdamAkiva) [reply Dockerfile]: There are still two typos in the file (lines 81 and 115 -> priviliges)

Comment (Meital-Kozhidov) [reply Dockerfile]: Thanks, I added this to the settings and now it works. I fixed the typos

File: fe/nginx/templates/dashboard.conf.template (hunk 11 above)
Comment (Meital-Kozhidov) [reply fe/nginx/templates/dashboard.conf.template:14]: You're literally the opposite of retarded

File: devops/docker-compose.yml (hunk 15 above)
Comment (Meital-Kozhidov) [reply devops/docker-compose.yml:92]: but we're trying to say that when you do 'dashboard.local.co.il' we'll refer you to 'localhost' no? does that not make more sense?

Comment (AdamAkiva) [reply devops/docker-compose.yml:92]: I don't know, maybe I'm missing the point but it sounds to me like this note means that you should map:
```text
dashboard.local.co.il -> localhost
```
When it should be the other way. (But again, maybe I have tunnel vision on the specific wording)
**Note:** from my google search it seems that it also must be 127.0.0.1 instead of `localhost` but I didn't check it and I can't check it right now, if you want you may attempt to check it

Comment (Meital-Kozhidov) [reply devops/docker-compose.yml]: I checked, it must be 127.0.0.1. Fixed the comment according to what you wrote, I guess it does make more sense
